zallet/src/components/json_rpc/balance.rs

```diff
@@ -0,0 +1,293 @@
+                    .and_then(|prev_tx| prev_tx.transparent_bundle())
+                    .and_then(|bundle| bundle.vout.get(txin.prevout.n() as usize))
+                {
+                    if is_mine(wallet, txout)? {
+                        acc = acc + txout.value;
+                    }
+                }
+            }
+            Ok(acc)
+        }
+    }
+}
+
+/// Equivalent to [`CWalletTx::GetCredit`] in `zcashd`.
+///
+/// [`CWalletTx::GetCredit`]: https://github.com/zcash/zcash/blob/2352fbc1ed650ac4369006bea11f7f20ee046b84/src/wallet/wallet.cpp#L4853
+pub(super) async fn wtx_get_credit(
+    wallet: &DbConnection,
+    chain: &FetchServiceSubscriber,
+    tx: &Transaction,
+    as_of_height: Option<BlockHeight>,
+    is_mine: impl Fn(&DbConnection, &TxOut) -> Result<bool, SqliteClientError>,
+) -> Result<Option<Zatoshis>, SqliteClientError> {
+    match tx.transparent_bundle() {
+        None => Ok(Some(Zatoshis::ZERO)),
+        // Must wait until coinbase is safely deep enough in the chain before valuing it.
+        Some(bundle)
+            if bundle.is_coinbase()
+                && wtx_get_blocks_to_maturity(wallet, chain, tx, as_of_height).await? > 0 =>
+        {
+            Ok(Some(Zatoshis::ZERO))
+        }
+        // Equivalent to `CWallet::GetCredit(CTransaction)` in `zcashd`.
+        Some(bundle) => {
+            let mut acc = Some(Zatoshis::ZERO);
+            for txout in &bundle.vout {
+                // Equivalent to `CWallet::GetCredit(CTxOut)` in `zcashd`.
+                if is_mine(wallet, txout)? {
+                    acc = acc + txout.value;
+                }
+            }
+            Ok(acc)
+        }
+    }
+}
+
+/// Equivalent to [`CWalletTx::IsFromMe`] in `zcashd`.
+///
+/// [`CWalletTx::IsFromMe`]: https://github.com/zcash/zcash/blob/2352fbc1ed650ac4369006bea11f7f20ee046b84/src/wallet/wallet.cpp#L4967
+pub(super) fn wtx_is_from_me(
+    wallet: &DbConnection,
+    tx: &Transaction,
+    is_mine: impl Fn(&DbConnection, &TxOut) -> Result<bool, SqliteClientError>,
+) -> Result<bool, SqliteClientError> {
+    if wtx_get_debit(wallet, tx, is_mine)?.ok_or_else(|| {
+        SqliteClientError::BalanceError(zcash_protocol::value::BalanceError::Overflow)
+    })? > Zatoshis::ZERO
+    {
+        return Ok(true);
+    }
+
+    wallet.with_raw(|conn| {
+        if let Some(bundle) = tx.sapling_bundle() {
+            let mut stmt_note_exists = conn.prepare(
+                "SELECT EXISTS(
+                    SELECT 1
+                    FROM sapling_received_notes
+                    WHERE nf = :nf
+                )",
+            )?;
+
+            for spend in bundle.shielded_spends() {
+                if stmt_note_exists
+                    .query_row(named_params! {":nf": spend.nullifier().0}, |row| row.get(0))?
+                {
+                    return Ok(true);
+                }
+            }
+        }
+
+        // TODO: Fix bug in `zcashd` where we forgot to add Orchard here.
```

Comment:
Fixed in https://github.com/zcash/zcash/pull/7035.

Reply:
Removed TODO at https://github.com/zcash/wallet/pull/36/commits/ac1011a286ce986c173e4f2b933e27b4bd038410